python/tests/ibd.py

```diff
@@ -0,0 +1,485 @@
+                if e.id < self.ts.num_edges - 1:
+                    e = next(mygen)
+                    continue
+                else:
+                    e = None
+
+            # Create a list of SegmentList objects, SegL, bookended by SegL_head and
+            # SegL_tail. Each SegmentList corresponds to a segment s in S, and
+            # contains all segments of samples that descend from s.
+            while S_head is not None:
+                u = S_head.node
+                list_to_add = SegmentList()
+                if u in self.samples:
+                    list_to_add + Segment(S_head.left, S_head.right, u)
+                else:
+                    s = self.A[u].head
+                    while s is not None:
+                        intvl = (max(S_head.left, s.left), min(S_head.right, s.right))
+                        if intvl[1] - intvl[0] > 0:
+                            list_to_add + Segment(intvl[0], intvl[1], s.node)
+                        s = s.next
+                self.add_to_chain(list_to_add)
+                S_head = S_head.next
+
+            # If we wanted to squash segments, we'd do it here.
+
+            # Use the info in SegL to find new ibd segments that descend from the
+            # parent node currently being processed.
+            if self.SegL_head.next is not None or self.current_parent in self.samples:
+                nodes_to_remove = self.calculate_ibd_segs()
+
+                # g. Remove elements of the list of ancestral segments A if they
+                # are no longer needed. (Memory-pruning step)
+                for n in nodes_to_remove:
+                    if self.current_parent in self.parent_list[n]:
+                        self.parent_list[n].remove(self.current_parent)
+                    if len(self.parent_list[n]) == 0:
+                        self.A[n] = []
+
+            # Create the list of ancestral segments, A, descending from current node.
+            # Concatenate all of the segment lists in SegL into a single list, and
+            # store in A. This is a list of all sample segments descending from
+            # the currently processed node.
+            seglist = self.SegL_head
+            while seglist is not None:
+                self.A[self.current_parent] + seglist
+                seglist = seglist.next
+
+            # Clear the temporary segment lists.
+            self.SegL_head = SegmentList()
+            self.SegL_tail = SegmentList()
+
+        return self.ibd_segments
+
+    def calculate_ibd_segs(self):
+        """
+        Calculates all new IBD segments found at the current iteration of the
+        algorithm. Returns information about nodes processed in this step, which
+        allows memory to be cleared as the procedure runs.
+        """
+
+        list0 = self.SegL_head
+        # Iterate through all pairs of segment lists.
+        while list0.next is not None:
+            list1 = list0.next
+            while list1 is not None:
```

Comment:
This looping over all pairs of lists is `N^2`, isn't it? Have you thought about how to avoid that?

Reply:
Thankfully not I think (I hope!)! In a 'normal' tree seq there will actually be only one iteration of this loop -- there will only be 'iterations' when the currently processed node is polytomous (?). There is a separate list for each newly coalescing set of segments underneath the node. Say for example you're processing a node `n` and it is parent to 3 edges in the tree sequence. Then there will be 3 lists in the list of lists, L1 L2 and L3 say, and this loop will iterate over (L1, L2), (L1, L3), (L2, L3).

Reply:
I have to say though, I do find it tricky to think about exactly *how* some of these procedures scale, though I can see that they are less than N^2. It would be good to spend more time pondering this so that in the manuscript we can give an explicit scaling factor. 

The  'loopiest' part of this algorithm is the looping over segments that happens inside this list loop you've highlighted here. At worst, if you were at the root of the trees, you could be comparing coalescing sets that each have n/2 segments in there, so a scaling factor of n^2/4. However, I don't think there's any way you could calculate IBD so far up a tree without doing that. 

A more sensible (though perhaps complicated?) way to assess how this scales would be to derive a scaling factor wrt IBD found at some depth in the tree.

Reply:
> At worst, if you were at the root of the trees, you could be comparing coalescing sets that each have n/2 segments in there, so a scaling factor of n^2/4

Right, good point - if we want pairwise IBD then we will output O(N^2) things in this case, no getting around it.